File: extras/rhubarb-for-spine/rhubarb-for-spine/AnimationFileModel.cs
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace rhubarb_for_spine {
	public class AnimationFileModel : ModelBase {
		private readonly string animationFilePath;
		private readonly SemaphoreSlim semaphore;
		private readonly JObject json;

		private string _mouthSlot;
		private MouthNaming _mouthNaming;
		private IReadOnlyCollection<MouthShape> _mouthShapes;
		private string _mouthShapesDisplayString;
		private BindingList<AudioFileModel> _audioFileModels;

		public AnimationFileModel(string animationFilePath, SemaphoreSlim semaphore) {
			this.animationFilePath = animationFilePath;
			this.semaphore = semaphore;
			json = SpineJson.ReadJson(animationFilePath);
			SpineJson.ValidateJson(json, AnimationFileDirectory);

			Slots = SpineJson.GetSlots(json);
			MouthSlot = SpineJson.GuessMouthSlot(Slots);
			var audioFileModels = SpineJson.GetAudioEvents(json)
				.Select(CreateAudioFileModel)
				.ToList();
			AudioFileModels = new BindingList<AudioFileModel>(audioFileModels);
		}

		public IReadOnlyCollection<string> Slots { get; }

		public string MouthSlot {
			get { return _mouthSlot; }
			set {
				_mouthSlot = value;
				MouthNaming = GetMouthNaming();
				MouthShapes = GetMouthShapes();
				OnPropertyChanged(nameof(MouthSlot));
			}
		}

		public MouthNaming MouthNaming {
			get { return _mouthNaming; }
			private set {
				_mouthNaming = value;
				OnPropertyChanged(nameof(MouthNaming));
			}
		}

		public IReadOnlyCollection<MouthShape> MouthShapes {
			get { return _mouthShapes; }
			private set {
				_mouthShapes = value;
				MouthShapesDisplayString = GetMouthShapesDisplayString();
				OnPropertyChanged(nameof(MouthShapes));
			}
		}

		public string MouthShapesDisplayString {
			get { return _mouthShapesDisplayString; }
			set {
				_mouthShapesDisplayString = value;
				SetError(nameof(MouthShapesDisplayString), GetMouthShapesDisplayStringError());
				OnPropertyChanged(nameof(MouthShapesDisplayString));
			}
		}

		public BindingList<AudioFileModel> AudioFileModels {
			get { return _audioFileModels; }
			set {
				_audioFileModels = value;
				if (!_audioFileModels.Any()) {
					SetError(nameof(AudioFileModels), "The JSON file doesn't contain any audio events.");
				}
				OnPropertyChanged(nameof(AudioFileModels));
			}
		}

		private string AnimationFileDirectory => Path.GetDirectoryName(animationFilePath);

		private MouthNaming GetMouthNaming() {
			IReadOnlyCollection<string> mouthNames = SpineJson.GetSlotAttachmentNames(json, _mouthSlot);
			return MouthNaming.Guess(mouthNames);
		}

		private List<MouthShape> GetMouthShapes() {
			IReadOnlyCollection<string> mouthNames = SpineJson.GetSlotAttachmentNames(json, _mouthSlot);
			return rhubarb_for_spine.MouthShapes.All
				.Where(shape => mouthNames.Contains(MouthNaming.GetName(shape)))
				.ToList();
		}

		private string GetMouthShapesDisplayString() {
			return MouthShapes
				.Select(shape => shape.ToString())
				.Join(", ");
		}

		private string GetMouthShapesDisplayStringError() {
			var basicMouthShapes = rhubarb_for_spine.MouthShapes.Basic;
			var missingBasicShapes = basicMouthShapes
				.Where(shape => !MouthShapes.Contains(shape))
				.ToList();
			if (!missingBasicShapes.Any()) return null;

			var result = new StringBuilder();
			string missingString = string.Join(", ", missingBasicShapes);
			result.AppendLine(missingBasicShapes.Count > 1
				? $"Mouth shapes {missingString} are missing."
				: $"Mouth shape {missingString} is missing.");
			MouthShape first = basicMouthShapes.First();
			MouthShape last = basicMouthShapes.Last();
			result.Append($"At least the basic mouth shapes {first}-{last} need corresponding image attachments.");
			return result.ToString();
		}

		private AudioFileModel CreateAudioFileModel(SpineJson.AudioEvent audioEvent) {
			string audioDirectory = SpineJson.GetAudioDirectory(json, AnimationFileDirectory);
			string filePath = Path.Combine(audioDirectory, audioEvent.RelativeAudioFilePath);
			bool animatedPreviously = SpineJson.HasAnimation(json, GetAnimationName(audioEvent.Name));
			var extendedMouthShapes = MouthShapes.Where(shape => !rhubarb_for_spine.MouthShapes.IsBasic(shape));
			return new AudioFileModel(
				audioEvent.Name, filePath, audioEvent.RelativeAudioFilePath, audioEvent.Dialog,
				new HashSet<MouthShape>(extendedMouthShapes), animatedPreviously,
				semaphore, cues => AcceptAnimationResult(cues, audioEvent));
		}

		private string GetAnimationName(string audioEventName) {
			return $"say_{audioEventName}";
		}

		private void AcceptAnimationResult(
			IReadOnlyCollection<MouthCue> mouthCues, SpineJson.AudioEvent audioEvent
		) {
			string animationName = GetAnimationName(audioEvent.Name);
			SpineJson.CreateOrUpdateAnimation(
				json, mouthCues, audioEvent.Name, animationName, MouthSlot, MouthNaming);
			File.WriteAllText(animationFilePath, json.ToString(Formatting.Indented));
		}
	}
}

File: extras/rhubarb-for-spine/src/main/kotlin/com/rhubarb_lip_sync/rhubarb_for_spine/AnimationFileModel.kt
package com.rhubarb_lip_sync.rhubarb_for_spine

import javafx.beans.binding.BooleanBinding
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleListProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.collections.ObservableList
import java.nio.file.Path
import tornadofx.getValue
import tornadofx.observable
import tornadofx.setValue
import java.util.concurrent.ExecutorService

class AnimationFileModel(animationFilePath: Path, private val executor: ExecutorService) {
	val spineJson = SpineJson(animationFilePath)

	val slotsProperty = SimpleObjectProperty<ObservableList<String>>()
	var slots by slotsProperty
		private set

	val mouthSlotProperty: SimpleStringProperty = SimpleStringProperty().alsoListen {
		mouthNaming = if (mouthSlot != null)
			MouthNaming.guess(spineJson.getSlotAttachmentNames(mouthSlot))
		else null

		mouthShapes = if (mouthSlot != null) {
			val mouthNames = spineJson.getSlotAttachmentNames(mouthSlot)
			MouthShape.values().filter { mouthNames.contains(mouthNaming.getName(it)) }
		} else listOf()

		mouthSlotError = if (mouthSlot != null) null
		else "No slot with mouth drawings specified."
	}
	var mouthSlot by mouthSlotProperty

	val mouthSlotErrorProperty = SimpleStringProperty()
	var mouthSlotError by mouthSlotErrorProperty
		private set

	val mouthNamingProperty = SimpleObjectProperty<MouthNaming>()
	var mouthNaming by mouthNamingProperty
		private set

	val mouthShapesProperty = SimpleObjectProperty<List<MouthShape>>().alsoListen {
		mouthShapesError = getMouthShapesErrorString()
	}
	var mouthShapes by mouthShapesProperty
		private set

	val mouthShapesErrorProperty = SimpleStringProperty()
	var mouthShapesError by mouthShapesErrorProperty
		private set

	val audioFileModelsProperty = SimpleListProperty<AudioFileModel>(
		spineJson.audioEvents
			.map { event ->
				AudioFileModel(event, this, executor, { result -> saveAnimation(result, event.name) })
			}
			.observable()
	)
	val audioFileModels by audioFileModelsProperty

	val busyProperty = SimpleBooleanProperty().apply {
		bind(object : BooleanBinding() {
			init {
				for (audioFileModel in audioFileModels) {
					super.bind(audioFileModel.busyProperty)
				}
			}
			override fun computeValue(): Boolean {
				return audioFileModels.any { it.busy }
			}
		})
	}
	val busy by busyProperty

	val validProperty = SimpleBooleanProperty().apply {
		val errorProperties = arrayOf(mouthSlotErrorProperty, mouthShapesErrorProperty)
		bind(object : BooleanBinding() {
			init {
				super.bind(*errorProperties)
			}
			override fun computeValue(): Boolean {
				return errorProperties.all { it.value == null }
			}
		})
	}
	val valid by validProperty

	private fun saveAnimation(mouthCues: List<MouthCue>, audioEventName: String) {
		val animationName = getAnimationName(audioEventName)
		spineJson.createOrUpdateAnimation(mouthCues, audioEventName, animationName, mouthSlot, mouthNaming)
		spineJson.save()
	}

	private fun getAnimationName(audioEventName: String): String = "say_$audioEventName"

	init {
		slots = spineJson.slots.observable()
		mouthSlot = spineJson.guessMouthSlot()
	}

	private fun getMouthShapesErrorString(): String? {
		val missingBasicShapes = MouthShape.basicShapes
			.filter{ !mouthShapes.contains(it) }
		if (missingBasicShapes.isEmpty()) return null

		val result = StringBuilder()
		val missingShapesString = missingBasicShapes.joinToString()
		result.appendln(
			if (missingBasicShapes.count() > 1)
				"Mouth shapes $missingShapesString are missing."
			else
				"Mouth shape $missingShapesString is missing."
		)

		val first = MouthShape.basicShapes.first()
		val last = MouthShape.basicShapes.last()
		result.append("At least the basic mouth shapes $first-$last need corresponding image attachments.")
		return result.toString()
	}

}